lib/requestContextPlugin.js

```diff
@@ -1,17 +1,37 @@
 const fp = require('fastify-plugin')
 const { als } = require('asynchronous-local-storage')
+const { AsyncResource } = require('async_hooks')
 
 const requestContext = {
   get: als.get,
   set: als.set,
 }
 
+/**
+ * Monkey patches `.emit()` method of the given emitter, so
+ * that all event listeners are run in scope of the provided
+ * async resource.
+ */
+const wrapEmitter = (emitter, asyncResource) => {
+  const original = emitter.emit
+  emitter.emit = function (type, ...args) {
+    return asyncResource.runInAsyncScope(original, emitter, type, ...args)
+  }
+}
+
+const wrapHttpEmitters = (req, res) => {
+  const asyncResource = new AsyncResource('fastify-request-context')
+  wrapEmitter(req, asyncResource)
+  wrapEmitter(res, asyncResource)
+}
+
 function plugin(fastify, opts, next) {
   fastify.decorate('requestContext', requestContext)
   fastify.decorateRequest('requestContext', requestContext)
 
-  fastify.addHook('onRequest', (req, res, done) => {
+  fastify.addHook(opts.hook || 'onRequest', (req, res, done) => {
     als.runWith(() => {
+      wrapHttpEmitters(req.raw, res.raw)
```

Comment:
@puzpuzpuz In your example I see that you are using `reply.res` instead of `res.raw`. Any reason for that?

Reply:
IIRC that property is not available in Fastify v2.x which my plugin requires.